File: LeetCode/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> rpt = new HashMap<>();
        int left = 0;
        int right = 0;
        int maxLen = 0;
        
        for (char c : s.toCharArray()) {
            if (rpt.containsKey(c) && left <= rpt.get(c)) left = rpt.get(c) + 1;
            
            rpt.put(c, right);
            
            if (maxLen < right - left + 1) maxLen = right - left + 1;
            
            right++;
        }
        
        return maxLen;
    }
}

File: LeetCode/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var start: Int = 0
        var end: Int = 0
        var max: Int = 0
        val map: MutableMap<Char, Int> = mutableMapOf()
        
        s.forEach { ch ->
            if (map.containsKey(ch) && start <= map[ch]!!) {
                start = map[ch]!! + 1
            }
            
            map[ch] = end
            
            max = Math.max(max, end - start + 1)
            
            end++
        }
        
        return max
    }
}